server/app/services/grid_scheduler.rb

```diff
@@ -58,6 +58,7 @@ def active_deploys_within_stack?(service)
   # @param [GridService] service
   # @return [Boolean]
   def lagging_behind?(service)
+    # TODO: check service.revision vs service_instance.service_revision instead?
     return true if service.deployed_at && service.updated_at > service.deployed_at
```

Comment:
I think this might be the final remaining usage of `GridService.updated_at` in the actual orchestration logic. It might make sense to change to use the `service.revision` instead somehow, but I'm not entirely sure how that should work... the `service.deployed_at` is still used for the actual agent deployment RPCs.

Reply:
There's also something buggy going on here that has a chance of trigger spurious re-deploys... although I think I've seen the same with `kontena stack deploy` on master?

```
D, [2017-11-27T12:48:14.436244 #8] DEBUG -- GridScheduler: service development/interval/redis updated_at=2017-11-27 12:47:53 UTC > deployed_at=2017-11-27T12:47:53+00:00
I, [2017-11-27T12:48:14.436381 #8]  INFO -- GridScheduler: service development/interval/redis does have older versions running
```

Reply:
Looks like the `GridService.deployed_at` is set using [`Time.now.utc.to_s`](https://github.com/kontena/kontena/blob/9bd81a08b4f2eeb46f9dfacf4d7d31ec33cd245a/server/app/services/grid_service_deployer.rb#L48), which strips any sub-second precision. This is tricky to fix because the entire agent RPCs => `ServicePodInstance.deploy_rev/rev` also uses the same timestamp for comparisons...

What ends up happening is that the `GridServices::Update/Deploy` -> `grid_service.updated_at` and the `GridServiceDeployer` -> `grid_service.deployed_at` will happen within the same second, and the `deployed_at` with the truncated sub-second precision will be less-than the `updated_at` that does have sub-second precision:

```
D, [2017-11-27T16:29:36.910617 #8] DEBUG -- GridScheduler: service development/interval/redis updated_at=2017-11-27T16:28:28.123000Z > deployed_at=2017-11-27T16:28:28.000000+00:00
I, [2017-11-27T16:29:36.910721 #8]  INFO -- GridScheduler: service development/interval/redis does have older versions running
I, [2017-11-27T16:29:36.913126 #8]  INFO -- GridScheduler: rescheduling service development/interval/redis
```

Reply:
AFAIK this should still be okay as-is: the `service.updated_at > service.deployed_at` logic should still work the same in this PR as before, with the only significant difference being that the `GridServices::Deploy` now also bumps the `updated_at` timestamps, but that should be okay together with the fix for #3091.TODO

Changing this to use the `service.revision` instead would need RPC changes to add a new `GridServiceInstance.service_revision` that gets reported by the agent.

Looks like this `service.updated_at > service.deployed_at` logic was added as part of the major refactoring in #1873, and it effectively made service deploys happen automatically even without `kontena service deploy`. Not entirely sure if that's the right behavior, even?